src/value/value.rs

```diff
@@ -31,7 +34,133 @@ mod serde;
 pub type StdError = Box<dyn std::error::Error + Send + Sync + 'static>;
 
 /// The storage mapping for the `Object` variant.
-pub type ObjectMap = BTreeMap<KeyString, Value>;
+// pub type ObjectMap = BTreeMap<KeyString, Value>;
+
+#[derive(Eq, PartialEq, Hash, Debug, Clone, Default, Serialize, Deserialize)]
+pub struct ObjectMap(pub Arc<BTreeMap<KeyString, Value>>);
+
+impl ObjectMap {
+    #[must_use]
+    pub fn new() -> Self {
+        ObjectMap(Arc::new(BTreeMap::new()))
+    }
+
+    pub fn into_values(self) -> impl Iterator<Item = Value> {
+        Arc::unwrap_or_clone(self.0).into_values()
+    }
+
+    pub fn into_keys(self) -> impl Iterator<Item = KeyString> {
+        Arc::unwrap_or_clone(self.0).into_keys()
+    }
+}
+
+impl IntoIterator for ObjectMap {
+    type Item = (KeyString, Value);
+    type IntoIter = std::collections::btree_map::IntoIter<KeyString, Value>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Arc::unwrap_or_clone(self.0).into_iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a ObjectMap {
+    type Item = (&'a KeyString, &'a Value);
+    type IntoIter = std::collections::btree_map::Iter<'a, KeyString, Value>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
+}
+
+impl FromIterator<(KeyString, Value)> for ObjectMap {
+    fn from_iter<T: IntoIterator<Item = (KeyString, Value)>>(iter: T) -> Self {
+        Self(Arc::new(BTreeMap::from_iter(iter)))
+    }
+}
+
+impl From<BTreeMap<KeyString, Value>> for ObjectMap {
+    fn from(value: BTreeMap<KeyString, Value>) -> Self {
+        Self(Arc::new(value))
+    }
+}
+
+impl<const N: usize> From<[(KeyString, Value); N]> for ObjectMap {
+    fn from(value: [(KeyString, Value); N]) -> Self {
+        Self(Arc::new(BTreeMap::from(value)))
+    }
+}
+
+impl Deref for ObjectMap {
+    type Target = BTreeMap<KeyString, Value>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for ObjectMap {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut *Arc::make_mut(&mut self.0)
+    }
+}
+
+#[derive(Eq, PartialEq, Hash, Debug, Clone, Default, Serialize, Deserialize)]
+pub struct ObjectArray(pub Arc<Vec<Value>>);
```

Comment:
I'm not personally a fan of this newtype name. The reasoning for `ObjectMap` was that it was an alias for the underlying `Map` type that is put into the `Object` variant. This is a wrapper for something going into the `Array` variant, so that part fits, but it has nothing to do with `Object`. Having said that, I'm having trouble coming up with a good parallel. `ArrayVec` is a tad redundant, and `ValueArray` isn't obvious. Thoughts?